blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,1066 @@
+    /// and push it to the pending phashmap
+    async fn add_helper(&mut self, msg: SignedMessage) -> Result<(), Error> {
+        add_helper(
+            self.api.clone(),
+            self.bls_sig_cache.clone(),
+            self.pending.clone(),
+            msg,
+        )
+        .await
+    }
+
+    /// Get the sequence for a given address, return Error if there is a failure to retrieve sequence
+    pub async fn get_sequence(&self, addr: &Address) -> Result<u64, Error> {
+        let cur_t = self.cur_tipset.read().await;
+        let cur_ts = cur_t.clone();
+        drop(cur_t);
+
+        let sequence = self.get_state_sequence(addr, &cur_ts).await?;
+
+        let pending = self.pending.read().await;
+
+        let msgset = pending.get(addr);
+        match msgset {
+            Some(mset) => {
+                if sequence > mset.next_sequence {
+                    return Ok(sequence);
+                }
+                Ok(mset.next_sequence)
+            }
+            None => Ok(sequence),
+        }
+    }
+
+    /// Get the state of the base_sequence for a given address in cur_ts
+    async fn get_state_sequence(&self, addr: &Address, cur_ts: &Tipset) -> Result<u64, Error> {
+        let api = self.api.read().await;
+        let actor = api.state_get_actor(&addr, cur_ts)?;
+        let mut base_sequence = actor.sequence;
+        // TODO here lotus has a todo, so I guess we should eventually remove cur_ts from one
+        // of the params for this method and just use the chain head
+        let msgs = api.messages_for_tipset(cur_ts)?;
+        drop(api);
+
+        for m in msgs {
+            if m.from() == addr {
+                if m.sequence() != base_sequence {
+                    return Err(Error::Other("tipset has bad sequence ordering".to_string()));
+                }
+                base_sequence += 1;
+            }
+        }
+
+        Ok(base_sequence)
+    }
+
+    /// Get the state balance for the actor that corresponds to the supplied address and tipset,
+    /// if this actor does not exist, return an error
+    async fn get_state_balance(&mut self, addr: &Address, ts: &Tipset) -> Result<BigInt, Error> {
+        let api = self.api.read().await;
+        let actor = api.state_get_actor(&addr, &ts)?;
+        Ok(BigInt::from(actor.balance))
+    }
+
+    /// Remove a message given a sequence and address from the messagepool
+    pub async fn remove(&mut self, from: &Address, sequence: u64) -> Result<(), Error> {
+        remove(from, self.pending.clone(), sequence).await
+    }
+
+    /// Return a tuple that contains a vector of all signed messages and the current tipset for
+    /// self.
+    pub async fn pending(&self) -> Result<(Vec<SignedMessage>, Tipset), Error> {
+        let mut out: Vec<SignedMessage> = Vec::new();
+
+        let pending = self.pending.read().await;
+        let pending_hm = pending.clone();
+        drop(pending);
```

Comment:
same, but it might make sense to just not drop the locks so they get dropped at end of function (is there an invariant that breaks this?)

Reply:
I don't see any reason to keep a read lock until the end of this function

Reply:
Is there no race condition you can hit with that pending cache and current tipset though? Do they have no relation to eachother that this will never be an issue? Point I'm making is that the pending lock is released before you lock the current tipset mutex, so there may be a potential invariant that breaks the logic of the function because another task could modify in between this?

Reply:
ah true didn't see that, yeah makes sense to keep that read lock until the end of the function